### PR TITLE
Add test helper to allow injecting listeners into the event bus

### PR DESCRIPTION
**Pull request by waiting-for-dev**
**Description**

It's not uncommon for registered listeners to make calls to external
services. Injecting a more lightweight listener when you want to test
that the corresponding event is fired helps make the test suite faster
and self-contained. Example:

```ruby
Spree::Event.enable_test_interface
listener = ExternalServiceSubscriber.listeners('foo').first
external_service_called = false
mocked_listener = proc { external_service_called = true }

Spree::Event.performing_only(listener) do
  Spree::Event.inject_listeners(listener: mocked_listener) do
    code_firing_foo_event
  end
end

expect(external_service_called).to be(true)
```

The mapping between original listeners and the injected ones is bound to
the block so that more complex workflows are supported:

```ruby
Spree::Event.inject_listeners(listener: mocked_listener) do |mapping|
  Spree::Event.unsubscribe(mapping[listener])
end
```

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

**Comment by aldesantis** [comment]
@waiting-for-dev I think we might want to put this on hold for a sec—I'd like to explore the possible use cases a bit more. Instead of replacing the listener, I'd prefer to see people implement a test mode/test adapter for their external service API, so that they don't have to mock the entire subscriber call. Mocking is very elegant, but it can also lead to brittle tests, and I'm not sure if we want to push people in this direction.

**Comment by waiting-for-dev** [comment]
> @waiting-for-dev I think we might want to put this on hold for a sec—I'd like to explore the possible use cases a bit more. Instead of replacing the listener, I'd prefer to see people implement a test mode/test adapter for their external service API, so that they don't have to mock the entire subscriber call. Mocking is very elegant, but it can also lead to brittle tests, and I'm not sure if we want to push people in this direction.

Thanks again for your feedback, @aldesantis. Related to https://github.com/nebulab/solidus/pull/155#issuecomment-969913093, I wanted it to be part of the injection approach. If you only make assumptions on the fired event name, you lose the opportunity to test the listener behavior unless you actually execute it. But sometimes, you'd instead inject a mock that behaves exactly like the original one but substitute an external call or something. However, we can hold it if we think it's not part of how people will use it.

**Comment by aldesantis** [comment]
@waiting-for-dev what I meant is that you can already do something like the following, which is more code but also allows you to use a test adapter globally for that service, no matter whether it's called from a listener, a background job enqueued by that listener, etc.:

```ruby
# app/lib/external_service.rb
module ExternalService
  def self.config
    @config ||= Config.new
  end

  def self.configure
    yield config
  end
end

# app/lib/external_service/config.rb
class ExternalService::Config
  attr_accessor :adapter
end

# app/lib/external_service/standard_adapter.rb
class ExternalService::StandardAdapter
  def call
    # real implementation here
  end
end

# app/lib/external_service/test_adapter.rb
class ExternalService::TestAdapter
  attr_reader :called

  def call
    @called = true
  end
end

# config/initializers/external_service.rb
ExternalService.configure do |config|
  config.adapter = Rails.env.test? ? ExternalService::TestAdapter : ExternalService::StandardAdapter
end
```

A simpler but equally effective alternative would be to implement a `test_mode` in your service, which might be a better idea if the module is not _really_ adapter-based, but only has one real-world implementation. If you have a DI container, you can also just use that.

In other words, I'd like to push devs in the direction of writing better, more easily testable code without having to replace listeners in the event bus entirely, and of getting into the habit of separating their listener-specific code (which belongs in the listener) from their business logic (which belongs somewhere else). This means more code gets executed during the tests and the tests are more reliable as a result.

What do you think?

**Comment by waiting-for-dev** [comment]
That's an excellent point. I hadn't thought about it, but providing such flexibility could indeed be a push towards a poor design. I'm happy to put it on hold and revisit it later if needed. I am closing the issue for now!
